pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,298 @@
+	// +optional
+	Spec RequestPrioritySpec
+	// Current status of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status RequestPriorityStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RequestPriorityList is a list of RequestPriority objects.
+type RequestPriorityList struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// Items is a list of request-priorities.
+	Items []RequestPriority
+}
+
+type RequestPrioritySpec struct {
+	// Exempted defines whether the priority level is exempted or not.  There should be at most one exempt priority level.
+	// Being exempt means that requests of that priority are not subject to concurrency limits (and thus are never queued)
+	// and do not detract from the concurrency available for non-exempt requests. In a more sophisticated system, the
+	// exempt priority level would be the highest priority level. It's defaulted to false and can only be true in the
+	// system preset priorities.
+	// +optional
+	Exempted *bool
+	// AssuredConcurrencyShares is a number representing the weight by which the priority level shares the concurrency
+	// from the global limit.The concurrency limit of an apiserver is divided among the non-exempt priority levels in
+	// proportion to their assured concurrency shares. Basically this produces the assured concurrency value (ACV) for each non-exempt
+	// priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
+	AssuredConcurrencyShares int32
+	// Queues is a number of queues that belong to a non-exempt RequestPriority object. The queues exist independently
+	// at each apiserver.
+	Queues int32
+	// HandSize is a small positive number for applying shuffle sharding. When a request arrives at an apiserver the
+	// request flow identifier’s string pair is hashed and the hash value is used to shuffle the queue indices and deal
+	// a hand of size. If empty, the hand size will the be set to 1 at which  the priority drains the queues on a FIFO basis.
+	// +optional
+	HandSize int32
+	// QueueLengthLimit is a length limit applied to each queue belongs to the priority.
+	// +optional
+	QueueLengthLimit int32
+}
+
+// RequestPriorityStatusConditionType is a valid value for RequestPriorityStatusCondition.Type
+type RequestPriorityStatusConditionType string
+
+// RequestPriorityStatusConditionStatus is a valid value for RequestPriorityStatusCondition.Status
+type RequestPriorityStatusConditionStatus string
+
+// RequestPriorityStatus represents the current state of a request-priority.
+type RequestPriorityStatus struct {
+	// CurrentQueues is the current number of existing queues that belong to this non-exempt RequestPriority object.
+	CurrentQueues *int32
```

Comment:
I don't see how this can work given that each apiserver implements this independently?

Reply:
mmm. indeed. what do you think if we make it a `map[string]int32` where `string` is the advertised address of apiserver?

Reply:
That invites update conflicts, when a bunch of apiservers react concurrently to a change.

Why do we need this field at all?

Reply:
> when a bunch of apiservers react concurrently to a change.

not necessary refresh the status reactively. we can flush the status at a constant interval? my point is that we should put more descriptive information in the status, at least the status should reflect whether the api configuration changes is captured by each apiserver?

Reply:
Flushing status periodically is more complicated and can still result in collisions (it is not hard to imagine a set of apiservers starting at the same time).

We already have logging in apiservers for how they handle configuration.  I am not sure what benefit we get from putting such status in API objects.

Reply:
now i remove those fields from the status, but i'm still trying put more reasonable descriptive information into the status. the flowcontrol works independently, we need more ways to observe a global status of the system or the users won't know about the changes when they modified the configuration. more details should be exposed from the system.

Reply:
"Users" in general do not care about this, it is a feature of concern to administrators --- the sort of people who have access to server logs.